File: mobile/flutter_app/lib/transport/bluetooth/bluetooth_transport.dart
import 'dart:async';
import 'dart:typed_data';
import 'package:flutter/services.dart';
import '../transport.dart';
import '../../protocol/frame_decoder.dart';
import '../../protocol/frame_encoder.dart';
import 'bluetooth_platform.dart';

class BluetoothTransport implements Transport {
  @override
  final TransportType type = TransportType.bluetooth;
  
  final String address;
  bool _isConnected = false;
  
  final _dataController = StreamController<Uint8List>.broadcast();
  final _closeController = StreamController<void>.broadcast();
  final _errorController = StreamController<dynamic>.broadcast();
  
  final _decoder = FrameDecoder();
  StreamSubscription? _eventSubscription;

  BluetoothTransport(this.address) {
    _decoder.onFrame.listen((frame) {
      _dataController.add(frame);
    });
  }

  @override
  bool get isConnected => _isConnected;

  @override
  Future<void> connect() async {
    try {
      _isConnected = await BluetoothPlatform.connect(address);
      if (_isConnected) {
        _eventSubscription = BluetoothPlatform.onEvents.listen(
          (dynamic event) {
            if (event is Uint8List) {
              _decoder.push(event);
            }
          },
          onError: (error) {
            _handleDisconnect(error);
          },
          onDone: () {
            _handleDisconnect(null);
          },
        );
      }
    } catch (e) {
      _isConnected = false;
      throw Exception('Failed to connect to Bluetooth device: $e');
    }
  }

  @override
  Future<void> disconnect() async {
    await BluetoothPlatform.disconnect();
    _handleDisconnect(null);
  }

  @override
  Future<void> send(Uint8List data) async {
    if (!_isConnected) throw Exception('Bluetooth socket disconnected');
    final framed = FrameEncoder.encode(data);
    await BluetoothPlatform.send(framed);
  }

  void _handleDisconnect(dynamic error) {
    if (!_isConnected) return;
    _isConnected = false;
    _eventSubscription?.cancel();
    
    if (error != null) {
      _errorController.add(error);
    }
    _closeController.add(null);
  }

  @override
  Stream<Uint8List> get onData => _dataController.stream;

  @override
  Stream<void> get onClose => _closeController.stream;

  @override
  Stream<dynamic> get onError => _errorController.stream;

  void dispose() {
    _dataController.close();
    _closeController.close();
    _errorController.close();
    _decoder.dispose();
  }
}

File: agent/src/transport/bluetooth/bluetooth-transport.ts
import net from 'net';
import { Transport, TransportType } from '../transport';
import { FrameEncoder, FrameDecoder } from './bluetooth-framing';

export class BluetoothTransport implements Transport {
  public readonly type = TransportType.BLUETOOTH;
  
  private dataHandlers: Array<(data: Buffer | string) => void> = [];
  private closeHandlers: Array<() => void> = [];
  private errorHandlers: Array<(err: Error) => void> = [];
  private decoder = new FrameDecoder();

  constructor(private socket: net.Socket, public deviceId: string) {
    this.decoder.onFrame((frame) => {
      this.dataHandlers.forEach(h => h(frame));
    });

    this.socket.on('data', (data) => {
      const buffer = typeof data === 'string' ? Buffer.from(data) : data;
      this.decoder.push(buffer);
    });

    this.socket.on('close', () => {
      this.closeHandlers.forEach(h => h());
    });

    this.socket.on('error', (err) => {
      this.errorHandlers.forEach(h => h(err));
    });
  }

  public async connect(): Promise<void> {
    return Promise.resolve();
  }

  public async disconnect(): Promise<void> {
    this.socket.destroy();
    return Promise.resolve();
  }

  public async send(data: Buffer | string): Promise<void> {
    if (!this.isConnected()) throw new Error('Socket disconnected');
    
    return new Promise((resolve, reject) => {
      const framedData = FrameEncoder.encode(data);
      this.socket.write(framedData, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  public onData(handler: (data: Buffer | string) => void): void {
    this.dataHandlers.push(handler);
  }

  public onClose(handler: () => void): void {
    this.closeHandlers.push(handler);
  }

  public onError(handler: (err: Error) => void): void {
    this.errorHandlers.push(handler);
  }

  public isConnected(): boolean {
    return !this.socket.destroyed && this.socket.writable;
  }
}
